pkg/distributor/ha_tracker.go

```diff
@@ -242,18 +243,21 @@ func replicasNotMatchError(replica, elected string) error {
 	return httpgrpc.Errorf(http.StatusAccepted, "replicas did not mach, rejecting sample: replica=%s, elected=%s", replica, elected)
 }
 
-// Modifies the labels parameter in place, removing labels that match
-// the replica or cluster label and returning their values. Returns an error
-// if we find one but not both of the labels.
+// Searches for the values of replicaLabel and clusterLabel
+// returning their values. Returns an error if we find one but not both of the labels.
 func findHALabels(replicaLabel, clusterLabel string, labels []client.LabelAdapter) (string, string) {
 	var cluster, replica string
 	var pair client.LabelAdapter
 
-	for _, pair = range labels {
+	for i := 0; i < len(labels) && (replica == "" || cluster == ""); i++ {
```

Comment:
The change set to this loop looks overcomplicated to me, just to break it earlier. I would suggest rolling back to the original implementation and simply switch `if pair.Name == clusterLabel {` to `if clusterLabel != "" && pair.Name == clusterLabel {`. I believe the performance you get breaking the loop earlier doesn't look justify making the loop way harder to read.

Reply:
see below comment